### PR TITLE
Egg incubation

### PR DESCRIPTION
**Pull request by reddivision**
Short Description: 
Egg incubation worker. Works with old protos and _should_ work with new protos (didn't get to test extensively there). 

There's definitely room for improvement once centralized Inventory class is implemented. 

We can also load this worker as an instance object of the bot for better tracking, or change the logic to utilize class variables for persistence, but I just stuck with existing patterns for workers for now.

Fixes:
- Not properly caring for pokemon eggs

### CODE REVIEW

**Comment by nburn42** [comment]
looks like egg_incubation_worker.py has a mix of tabs and spaces

**Comment by reddivision** [comment]
@nburn42 that's what I get for using vim. should be better now.

**Comment by reddivision** [comment]
nope, it's not... one sec lol

File: pokemongo_bot/__init__.py [hunk 1]
```diff
@@ -100,7 +100,10 @@ def update_web_location(self, cells=[], lat=None, lng=None, alt=None):
                                                      gym_latitude=fort.get('latitude'),
                                                      gym_longitude=fort.get('longitude'))
                             response_gym_details = self.api.call()
-                            fort['gym_details'] = response_gym_details['responses']['GET_GYM_DETAILS']
+                            try:
+                                fort['gym_details'] = response_gym_details['responses']['GET_GYM_DETAILS']
+                            except:
+                                print fort
```
Comment (bstpierr) [inline pokemongo_bot/__init__.py]: Was this for debugging?

File: pokemongo_bot/cell_workers/egg_incubation_worker.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    def work(self):
+        check_incubation = False
+        self._check_inventory() # get km, incubators, and incubating pkmn
+        self._calculate_remaining_km()
+        if len(self.remaining_km) and min(self.remaining_km)<=0:
+            logger.log('[#] Checking for hatched eggs...')
+            check_incubation = self._check_hatched_eggs()
+            sleep(10)
+            self.bot.latest_inventory = None
+            self._check_inventory() # incubators don't release til after check hatched eggs call
```
Comment (bstpierr) [inline pokemongo_bot/cell_workers/egg_incubation_worker.py]: The heartbeat current calls the hatched egg check. Not sure how that might affect your code.

**Comment by nburn42** [comment]
This might not be useful because I merged these changes into my master branch not the dev branch.
I got it working with a few tweaks.

Primarily this line threw an exception
line 93 of pokemongo_bot/cell_workers/egg_incubation_worker.py
`elif inc['uses_remaining']>0:`

There was no 'uses_remaining" for me. This is probably a nonissue and some change in the dev branch adds this information to the item.

Cheers

**Comment by mrseeker** [comment]
From what I can see from testing is that it won't incubate eggs using the free unlimited incubator, unless I am missing something?

**Comment by koppa** [comment]
Still has some problems:

```
[22:26:45] [#] Checking for hatched eggs...
[22:26:45] [d] Get hatched eggs call result: {'responses': {'GET_HATCHED_EGGS': {'stardust_awarded': [1203, 1568, 1600], 'candy_awarded': [16, 20, 21], 'experience_awarded': [500, 1000, 500], 'success': True, 'pokemon_id': [3091093573105956752L, 1071922182428598967L, 6596723275451026452L]}}, 'status_code': 1, 'auth_ticket': {'expire_timestamp_ms': 1469653005540L, 'start': 'kRDnzCLdDjPc81JoSi6rbTONsTqrf/f99qpQtxKeFnYp/bprRGLn8uce8q8O4mBGwas4i7gOYwyPE045bZi/jg==', 'end': 'H001rvewoj3yHgdxelyrEA=='}, 'request_id': 8145806132888207460L}
[22:26:45] [#] Hatched eggs check call failed!
[22:26:54] [#] Applying incubators to eggs...
[22:26:54] [#] 3 incubators available, 5 eggs ready.
```

**Comment by koppa** [comment]
Another bug I found:

```
[23:10:46] [#] Hatched eggs check call failed!
Traceback (most recent call last):
  File "./pokecli.py", line 289, in <module>
    main()
  File "./pokecli.py", line 255, in main
    bot.take_step()
  File "/home/markus/poke/PokemonGo-Bot/pokemongo_bot/__init__.py", line 47, in take_step
    self.process_cells()
  File "/home/markus/poke/PokemonGo-Bot/pokemongo_bot/__init__.py", line 67, in process_cells
    "catchable_pokemons": catchable_pokemons}, location)
  File "/home/markus/poke/PokemonGo-Bot/pokemongo_bot/__init__.py", line 178, in work_on_cell
    worker.work()
  File "/home/markus/poke/PokemonGo-Bot/pokemongo_bot/cell_workers/egg_incubation_worker.py", line 34, in work
    logger.log('[#] Next egg hatches in {:.2n} km.'.format(min([km if km>=0 else 0 for km in self.remaining_km])),'yellow')
ValueError: Precision not allowed in integer format specifier
```

**Comment by reddivision** [comment]
@mrseeker you're right, it won't. For some reason the id for the unlimited incubator is nowhere to be found in the inventory response. I didn't catch this at first, but can't see much to do about it for the time being.

**Comment by reddivision** [comment]
@koppa, yes, I've got a fix for this and for applying incubators correctly in the pipeline (before was doing one at a time and freezing after application)

**Comment by Nihisil** [comment]
@reddivision "901" - isn't it an id for free unlimited incubator?

**Comment by reddivision** [comment]
It's the item_id, but there's another id that the api requires when calling the method to assign an incubator to a pokemon.

**Comment by Nihisil** [comment]
I see. But this code don't have any sense if it can't use unlimited egg incubator. I will try to investigate how to get a unlimited incubator id.

Also, there is a inventory cache implementation in dev. Can you merge it to your PR? https://github.com/PokemonGoF/PokemonGo-Bot/commit/3b51bddc5f11f6f54fd4348579a0e54134f7cea1

**Comment by reddivision** [comment]
you can have more than one incubator, so there is a current use for it based on what we have access to. I've already merged the inventory cache implementation. I'm closing this since it's outdated. I'll push the new one after I've verified fixes are worked.

File: pokemongo_bot/cell_workers/egg_incubation_worker.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+                    item = item['inventory_item_data']
+                    if 'egg_incubators' in item:
+                        incubators = item['egg_incubators']
+                        if isinstance(incubators, list):
+                            #new protos
+                            for inc in incubators:
+                                incubator_list.append(inc['egg_incubator'])
+                        else:
+                            #old protos
+                            incubator_list.append(incubators['egg_incubator'])
```
Comment (rguedes) [inline pokemongo_bot/cell_workers/egg_incubation_worker.py]: This line not work. Change to:
`for inc in incubators['egg_incubator']:
      incubator_list.append(inc)
`

File: pokemongo_bot/cell_workers/egg_incubation_worker.py [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+                        if 'is_egg' in pokemon and pokemon['is_egg']:
+                            egg_list.append(pokemon['id'])
+                    if 'player_stats' in item:
+                        self.km_walked = item['player_stats']['km_walked']
+            if len(incubator_list)>0:
+                for inc in incubator_list:
+                    if 'pokemon_id' in inc:
+                        self.incubating_eggs.append(inc['pokemon_id'])
+                        self.used_incubators.append(inc)
+                    elif inc['uses_remaining']>0:
```
Comment (rguedes) [inline pokemongo_bot/cell_workers/egg_incubation_worker.py]: Must be:
`elif 'uses_remaining' in inc and inc['uses_remaining'] > 0:`

**Pull request closed without merging**